### ISSUE TITLE
sharenfs options not documented properly

### ISSUE THREAD

**Issue by vozhyk-** [zfsonlinux/zfs#3860]
`zfs(8)` at `sharenfs` points to `exportfs(8)`, which points to `exports(5)`, which have descriptions of Linux export options. But ZFS uses the Solaris ones (which are documented at [[1]](https://docs.oracle.com/cd/E19455-01/806-0625/6j9vfilvs/index.html).
Also, `zfs(8)` doesn't mention the default options that ZFS uses ([[2]](https://github.com/zfsonlinux/zfs/blob/master/lib/libshare/nfs.c#L388)).

**Comment by vozhyk-** [comment]
Sorry, I should have read that more carefully.

**Comment by FransUrbo** [comment]
@vozhyk- Could you please close this then?

**Comment by vozhyk-** [comment]
It still says that if `sharenfs` != `on`, the `sharenfs` options will be used *instead* of the defaults (so a user might understand it as "I have to specify `no_subtree_check,no_root_squash` if I set `sharenfs` to something other than `on`"), which is not the actual behavior.

**Comment by FransUrbo** [comment]
That is correct. If you specify any option other than <code>on</code> in <code>sharenfs</code>, those will be used instead of the defaults. So you would have to specify <code> no_subtree_check,no_root_squash[,etc]</code>.

**Comment by vozhyk-** [comment]
If I read the code correctly (the [2] link is the 0th comment) that's not the case - the defaults (`no_subtree_check,no_root_squash`) work even when `sharenfs=<some_other_options>`.
Just tested in practice:
```
# zfs set sharenfs=rw,crossmnt hts/funtoo/usr/portage/distfiles hts/freebsd
## rsv is an alias for restarting services;
## run just in case zfs doesn't change the export options
# rsv nfs zfs-share
nfs                 | * Stopping NFS mountd ...                            [ ok ]
nfs                 | * Stopping NFS daemon ...                            [ ok ]
nfs                 | * Unexporting NFS directories ...                    [ ok ]
nfs                 | * Starting NFS mountd ...                            [ ok ]
nfs                 | * Starting NFS daemon ...                            [ ok ]
nfs                 | * Starting NFS smnotify ...                          [ ok ]
zfs-share           | * Unsharing ZFS filesystems  ...                     [ ok ]
zfs-share           | * Sharing ZFS filesystems  ...                       [ ok ]

# mount -t nfs localhost:/freebsd /mnt/h
# touch /mnt/h/test2
# ls -l /mnt/h/test2
-rw-r--r--  1 root root       0 Кст  4 20:46 test2
```
So at least `no_root_squash` is applied, even if it wasn't specified explicitly.

**Comment by FransUrbo** [comment]
Ah, yes! I rewrote this in #2790. Now when I think about it, that was one of the many problems with the <code>sharenfs</code> code..

Unfortunately, there's little chance of getting this fixed the next couple of years...

**Comment by vozhyk-** [comment]
The behavior here is good for me (the defaults can be overridden with `root_squash,subtree_check` if needed) (not talking about per-host options, which I didn't try).

**Comment by optiz0r** [comment]
I just ran into an issue where setting `sharenfs=sec=krb5` did not result in the filesystem being shared. As per https://github.com/zfsonlinux/zfs/issues/3738#issuecomment-139488424 setting `sharenfs=rw,sec=krb5` did share the filesystem, but this requirement to prefix the share options with `ro` or `rw` is not covered in the manpage (at least for 0.7.6)

**Comment by siebenmann** [comment]
Is there any chance of getting the zfs(8) manpage documentation of sharenfs fixed any time soon? This issue has been open for years, and the documentation remains extremely misleading, which is frustrating. I suggest that the minimal update would be to note that sharenfs must use Illumos option names with the exception that '@<something>' as a host/netgroup/etc name will be accepted (I think that's accurate based on reading the source), and that not all Illumos options are accepted.

(If such a change would be accepted, I'd be happy to write up a tentative patch.)

**Comment by behlendorf** [comment]
@siebenmann absolutely, by all means please open a PR with the clarifcations.

**Comment by gdevenyi** [comment]
Following up on this unclear documentation, is it possible to actually set different export options for different IPs? If so, what's the order? Options then IP?

**Comment by hvisage** [comment]
Have the same issue, need different options different IPs, but seems to be better to do it via /etc/exports ??

**Comment by awehrfritz** [comment]
@FransUrbo and @behlendorf: has there been any progress on this issue recently? It seems there was an attempt to improve this situation in PR #2790, that was never merged. Any update on this would be much appreciated.

**Comment by rmbleeker** [comment]
Regarding questions about setting different options per IP, I've found that the following works:
```
sharenfs="root_squash,rw=192.168.1.216,ro=192.168.1.0/24"
```
This shares the dataset read-write to one host and read-only to all other hosts in the network. All connections are root_squashed. If I try:
```
sharenfs="root_squash,rw=192.168.1.216,no_root_squash,ro=192.168.1.0/24"
```
Then all shares end up with the `no_root_squash` property, leading me to believe that NFS sharing through ZFS can only differentiate between read-write and read-only access on a single share.

Also, the @ symbol is only useful when using partial ip addresses. In other words, `192.168.1.0/24` and `@192.168.1` are equivalent.
